### PR TITLE
Add Tensor Tracing to inline-dynamic lowerings

### LINKED ISSUE #14970: `iree-hal-loader-conversion` fails with conversion error on `inline-dynamic`
### What happened?

For the following test case:
```
func.func @main(%arg1: tensor<41xi64>, %144: tensor<10x128xf32>) -> tensor<1x45x128xf32> {
    %145 = "stablehlo.reshape"(%arg1) : (tensor<41xi64>) -> tensor<1x41xi64>
    %146 = "stablehlo.constant"() {value = dense<9> : tensor<i64>} : () -> tensor<i64>
    %147 = "stablehlo.pad"(%145, %146) {edge_padding_high = dense<[0, 2]> : tensor<2xi64>, edge_padding_low = dense<[0, 2]> : tensor<2xi64>, interior_padding = dense<0> : tensor<2xi64>} : (tensor<1x41xi64>, tensor<i64>) -> tensor<1x45xi64>
    %148 = "stablehlo.gather"(%144, %147) {dimension_numbers = #stablehlo.gather<offset_dims = [2], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 2>, indices_are_sorted = false, slice_sizes = dense<[1, 128]> : tensor<2xi64>} : (tensor<10x128xf32>, tensor<1x45xi64>) -> tensor<1x45x128xf32>
    return %148 : tensor<1x45x128xf32>
}
```

When compiled using the following command (HEAD: bc2a32d47bd2c32571b9f830da4a71bb1211d63a):

```
./iree-compile --iree-hal-target-backends=llvm-cpu --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu  --iree-llvmcpu-enable-microkernels=false --iree-codegen-llvm-disable-distribution=true --iree-llvmcpu-link-embedded=false --iree-vm-emit-polyglot-zip=true --iree-llvmcpu-debug-symbols=true --iree-vm-bytecode-module-strip-source-map=false --iree-llvmcpu-reassociate-fp-reductions --iree-input-demote-f64-to-f32=true --iree-input-demote-i64-to-i32=true --iree-llvmcpu-fail-on-out-of-bounds-stack-allocation=false  small_tester.mlir -o small_tester.vmfb --iree-execution-model=inline-dynamic --iree-flow-trace-dispatch-tensors  --mlir-disable-threading
```

Is giving the following errors:

```
small_tester.mlir:6:12: error: failed to materialize conversion for result #0 of operation 'stream.tensor.export' that remained live after conversion
    %148 = "stablehlo.gather"(%144, %147) {dimension_numbers = #stablehlo.gather<offset_dims = [2], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 2>, indices_are_sorted = false, slice_sizes = dense<[1, 128]> : tensor<2xi64>} : (tensor<10x128xf32>, tensor<1x45xi64>) -> tensor<1x45x128xf32>
           ^
small_tester.mlir:2:1: note: called from
func.func @main(%arg1: tensor<41xi64>, %144: tensor<10x128xf32>) -> tensor<1x45x128xf32> {
^
small_tester.mlir:6:12: note: see current operation: %31 = "stream.tensor.export"(%24, %2) {source_encoding = tensor<1x45xi32>} : (!stream.resource<external>, index) -> tensor<1x45xi32>
    %148 = "stablehlo.gather"(%144, %147) {dimension_numbers = #stablehlo.gather<offset_dims = [2], collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 2>, indices_are_sorted = false, slice_sizes = dense<[1, 128]> : tensor<2xi64>} : (tensor<10x128xf32>, tensor<1x45xi64>) -> tensor<1x45x128xf32>
           ^
small_tester.mlir:6:12: note: see existing live user here: "hal_inline.buffer_view.trace"(%31, %38) {key = "main_dispatch_0::main_dispatch_0_generic_1x45x128_f32 inputs"} : (tensor<1x45xi32>, tensor<10x128xf32>) -> ()
small_tester.mlir:0:0: error: conversion to the hal_inline + hal_loader dialects failed
small_tester.mlir:0:0: note: see current operation:
...
```

The enabling or disabling demotion from i64->i32 or f64->f32 doesn't change the behavior.

This issue only happens when `--iree-flow-trace-dispatch-tensors` and `--iree-execution-model=inline-dynamic` is used together.

### Steps to reproduce your issue

Please see what happened section.

### What component(s) does this issue relate to?

_No response_

### Version information

HEAD: bc2a32d47bd2c32571b9f830da4a71bb1211d63a

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rsuderman**
The TypeConverter in StreamToHALLoader was missing the right type conversion causing failures when lowering to inlined HAL. Updated for the correct type conversion.

### CODE REVIEW

File: compiler/src/iree/compiler/Modules/HAL/Loader/Transforms/Conversion.cpp [hunk 1]
```diff
@@ -49,11 +50,19 @@ class ConversionPass : public ConversionBase<ConversionPass> {
         mlir::func::FuncDialect, mlir::scf::SCFDialect,
         mlir::arith::ArithDialect, mlir::affine::AffineDialect>();
 
-    TypeConverter typeConverter;
+    SmallVector<const HALConversionDialectInterface *> conversionInterfaces;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Modules/HAL/Loader/Transforms/Conversion.cpp:53]: please make the same change in HAL/Inline/Transforms/Conversion.cpp

File: compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir [hunk 2]
```diff
@@ -89,3 +89,29 @@ func.func @cmdDispatch(%buffer0: !stream.resource<transient>, %buffer0_size: ind
...
+  %tensor = stream.tensor.export %arg0 : tensor<1x45xi32> in !stream.resource<external>{%c180} -> tensor<1x45xi32>
+
+  // CHECK: hal_inline.buffer_view.trace %[[VIEW]] : !hal.buffer_view attributes {key = "whatevs"}
+  stream.tensor.trace {key = "whatevs"} %tensor : tensor<1x45xi32>
+
+  // CHECK: %[[VIEW:.+]] = hal_inline.buffer_view.create buffer(%[[ARG0]] : !hal.buffer)[%[[C0]], %[[C180]]] shape([%[[C1]], %[[C45]]]) type(%[[C268435488]]) encoding(%[[C1_i32]]) : !hal.buffer_view
+  %buffer = stream.tensor.export %arg0 : tensor<1x45xi32> in !stream.resource<external>{%c180} -> !hal.buffer_view
+
+  // CHECK: return %[[VIEW]] : !hal.buffer_view
+  return %buffer : !hal.buffer_view
```
Comment (benvanik) [inline compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir]: you can remove this extra IR as it's not adding additional testing vs the first one of the same export
```suggestion
```

File: compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir [hunk 3]
```diff
@@ -89,3 +89,29 @@ func.func @cmdDispatch(%buffer0: !stream.resource<transient>, %buffer0_size: ind
...
+func.func @trace_tensor(%arg0: !stream.resource<external>) -> !hal.buffer_view {
+  // CHECK: %[[C180:.+]] = arith.constant 180 : index
+  %c180 = arith.constant 180 : index
+
+  // CHECK: %[[C1:.+]] = arith.constant 1 : index
+  // CHECK: %[[C45:.+]] = arith.constant 45 : index
+  // CHECK: %[[C0:.+]] = arith.constant 0 : index
+  // CHECK: %[[C268435488:.+]] = arith.constant 268435488 : i32
+  // CHECK: %[[C1_i32:.+]] = arith.constant 1 : i32
+  // CHECK: %[[VIEW:.+]] = hal_inline.buffer_view.create buffer(%arg0 : !hal.buffer)[%[[C0]], %[[C180]]] shape([%[[C1]], %[[C45]]]) type(%[[C268435488]]) encoding(%[[C1_i32]]) : !hal.buffer_view
```
Comment (benvanik) [inline compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir]:
```suggestion
  // CHECK: %[[VIEW:.+]] = hal_inline.buffer_view.create buffer(%[[ARG0]] : !hal.buffer)[%[[C0]], %[[C180]]] shape([%[[C1]], %[[C45]]]) type(%[[C268435488]]) encoding(%[[C1_i32]]) : !hal.buffer_view
```

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir [hunk 4]
```diff
@@ -109,9 +109,4 @@ func.func @trace_tensor(%arg0: !stream.resource<external>) -> !hal.buffer_view {
   // CHECK: hal_inline.buffer_view.trace %[[VIEW]] : !hal.buffer_view attributes {key = "whatevs"}
   stream.tensor.trace {key = "whatevs"} %tensor : tensor<1x45xi32>
 
-  // CHECK: %[[VIEW:.+]] = hal_inline.buffer_view.create buffer(%[[ARG0]] : !hal.buffer)[%[[C0]], %[[C180]]] shape([%[[C1]], %[[C45]]]) type(%[[C268435488]]) encoding(%[[C1_i32]]) : !hal.buffer_view
```
Comment (benvanik) [inline compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir]: you'll have to remove the func return type too (sorry, missed that in the diff)

Comment (rsuderman) [reply compiler/src/iree/compiler/Modules/HAL/Loader/Conversion/StreamToHALLoader/test/cmd_ops.mlir]: Fixed.

File: compiler/src/iree/compiler/Modules/HAL/Loader/Transforms/Conversion.cpp (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Modules/HAL/Loader/Transforms/Conversion.cpp:53]: Implemented.

**Review by benvanik — APPROVED** [review]

**Pull request merged**